pkg/app/api/grpcapi/piped_api.go

```diff
@@ -1206,13 +1210,44 @@ func (a *PipedAPI) InChainDeploymentPlannable(ctx context.Context, req *pipedser
 		return nil, status.Error(codes.InvalidArgument, "invalid deployment with chain block index provided")
 	}
 
-	prevBlockSuccessfullyCompleted, err := dc.IsSuccessfullyCompletedBlock(req.DeploymentChainBlockIndex - 1)
-	if err != nil {
-		return nil, status.Error(codes.InvalidArgument, "unable to process the previous block of this deployment in chain")
+	previousBlock := dc.Blocks[req.DeploymentChainBlockIndex-1]
+	isPreviousBlockFinished := previousBlock.IsCompleted()
+
+	// If the previous block has not finished yet, should not plan this deployment to run.
+	if !isPreviousBlockFinished {
+		return &pipedservice.InChainDeploymentPlannableResponse{
+			Plannable: false,
+		}, nil
+	}
+
+	var plannable bool
+	switch previousBlock.Status {
+	case model.ChainBlockStatus_DEPLOYMENT_BLOCK_SUCCESS:
+		plannable = true
+	case model.ChainBlockStatus_DEPLOYMENT_BLOCK_FAILURE, model.ChainBlockStatus_DEPLOYMENT_BLOCK_CANCELLED:
+		plannable = false
+		// Send CANCEL_DEPLOYMENT command so that the in-charge piped will be aware and stop the deployment.
+		if err = addCommand(ctx, a.commandStore, &model.Command{
+			Id:            uuid.New().String(),
+			PipedId:       pipedID,
+			ApplicationId: req.ApplicationId,
+			ProjectId:     dc.ProjectId,
+			DeploymentId:  req.DeploymentId,
+			Type:          model.Command_CANCEL_DEPLOYMENT,
+			Commander:     dc.Id,
+			CancelDeployment: &model.Command_CancelDeployment{
+				DeploymentId:  req.DeploymentId,
+				ForceRollback: true,
+			},
+		}, a.logger); err != nil {
+			return nil, status.Error(codes.Internal, "unable to send cancel deployment command to stop this deployment")
+		}
```

Comment:
Instead of sending a new command to stop all deployments behind, how about adding the reason why the deployment could not be plannable to the RPC response then Piped can stop them immediately.
I think that way is more reliable than more than one command could be added for the same deployment.

Reply:
I get your point 👍  That way's better, lets me adopt that 👍

Reply:
Sounds nicer 👍

Reply:
addressed by https://github.com/pipe-cd/pipe/pull/2866/commits/cefa70657bf2b511b08cf45a89e86c12b75a2a2b, PTAL 🙏